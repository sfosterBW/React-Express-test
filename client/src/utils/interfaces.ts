export interface Fruit {
  id: string
  name: string
  best: boolean
  description?: string
}

export type NewFruit = Omit<Fruit, 'id'>
