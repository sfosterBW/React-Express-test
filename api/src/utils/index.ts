import { IFruit, NewFruit } from './types'

/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment*/

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

export const toNewFruit = (object: any): NewFruit => {
  const newFruit: NewFruit = {
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return newFruit
}

const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

export const toFruit = (object: any): IFruit => {
  const fruit: IFruit = {
    ...object,
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return fruit
}
