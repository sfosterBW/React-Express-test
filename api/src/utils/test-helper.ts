import Fruit from '../models/fruit'

const initialFruits = [
  {
    name: 'apple',
    best: false,
    description: "They keep doctors away"
  },
  {
    name: 'orange',
    best: true
  }
]

const fruitsInDb = async () => {
  const fruits = await Fruit.find({})
  return fruits
}

const nonExistingId = async () => {
  const fruit = new Fruit({
    name: 'willdelete',
    best: false
  })

  await fruit.save()
  await fruit.remove()

  return fruit.id
}

export default { initialFruits, fruitsInDb, nonExistingId }
