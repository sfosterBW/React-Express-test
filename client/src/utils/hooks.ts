import { useState } from 'react'

export const useField = (
  label: string,
  name: string,
  type: string,
  initValue: string = ""
) => {

  const [value, setValue] = useState(initValue)

  const handleChange = (event: any): void => {
    if (type === "checkbox") {
      setValue(event.target.checked)
    } else {
      setValue(event.target.value)
    }
  }

  const reset = (resetValue = ""): void => {
    setValue(resetValue)
  }

  return {
    handleChange,
    label,
    name,
    reset,
    type,
    value
  }
}
