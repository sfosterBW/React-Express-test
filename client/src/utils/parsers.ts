import { Fruit } from './interfaces'

/* eslint-disable  @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`)
  }

  return best
}

export const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

const parseArray = (array: any): Array<unknown> => {
  if (!array || !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${String(array)}`)
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  const fruit: Fruit = {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }

  if (object.description) {
    fruit.description = parseDescription(object.description)
  }

  return fruit
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
