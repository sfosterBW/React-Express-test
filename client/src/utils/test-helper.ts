import { Fruit } from './interfaces'

export const emptyList: Fruit[] = []

export const fruit: Fruit = {
  id: "1",
  name:"apple",
  best:false,
  description: "I don't like doctors"
}

export const list: Fruit[] = [
  {
    id: "1",
    name: "apple",
    best: false,
    description: "I don't like doctors"
  },
  {
    id: "2",
    name: "banana",
    best: true
  }
]

export default { emptyList, fruit, list }
