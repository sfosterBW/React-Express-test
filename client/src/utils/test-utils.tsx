import React, { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from "redux-mock-store"
import thunk from "redux-thunk"

import { initState as initAlertState } from '../Reducers/alertReducer'
import { initState as initFruitState } from '../Reducers/fruitReducer'
import { initState as initModalState } from '../Reducers/modalReducer'
import { RootState } from '../Reducers/store'

export const rootInitialState = {
  alert: initAlertState,
  fruit: initFruitState,
  modal: initModalState
};

export const renderWithProviders = (
  ui: ReactNode,
  initialState: RootState = rootInitialState
) => {
  const store = configureStore([thunk])(initialState)
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    mockStore: store,
    getActionsTypes: () => store.getActions().map(a => a.type)
  }
}
