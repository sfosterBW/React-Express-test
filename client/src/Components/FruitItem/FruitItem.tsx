import React, { FC, useState } from 'react'
import { useDispatch } from 'react-redux'
import { Fruit } from '../../utils/interfaces'
import { openModal } from '../../Reducers/modalReducer'
import { removeFruit, updateFruit } from '../../Reducers/fruitReducer'
import styles from './FruitItem.module.css'

interface Props {
  active?: boolean
  fruit: Fruit
}

const FruitItem: FC<Props> = ({ active = false, fruit }) => {
  const dispatch = useDispatch()

  const [activeToggle, setActiveToggle] = useState<boolean>(active)
  
  const { id, name, best, description } = fruit

  return (
    <div className={styles.row} data-testid="item">
      <div
        className={styles.header}
        onClick={() => setActiveToggle(!activeToggle)}
      >
        <div className={styles.section}>
          <h3 className={styles.subtitle} data-testid="item-title">
            {name}
          </h3>
        </div>
      </div>
      <div
        className={styles.active}
        style={{
          height: activeToggle ? "calc(120px + 12vmin)" : "0",
          transition: "all 0.5s"
        }}
      >
        <div className={styles.sectionRow}>
          <h4 className={styles.subtitle}>
            Description
          </h4>
          <p className={styles.description} data-testid="item-description">
            {description ? `${description}` : "No description added yet"}
          </p>
        </div>
        <div className={styles.section}>
          <h4 className={styles.subtitle}>
            Best
          </h4>
          <input
            checked={best}
            className={styles.button}
            id={id}
            name={id}
            onChange={() => {
              dispatch(updateFruit({ ...fruit, best: !best }))
            }}
            type="checkbox"
            value={fruit.id}
          />
          <label data-testid="item-best" htmlFor={id}>
            {`${best}`}
          </label>
        </div>
        <div className={styles.section}>
          <h4 className={styles.subtitle}>
            Edit
          </h4>
          <button
            className={styles.button}
            data-testid="update-button"
            name="modal"
            onClick={() => dispatch(openModal(fruit))}
          >
            Update
          </button>
          <button
            className={styles.button}
            data-testid="remove-button"
            name="remove"
            onClick={() => dispatch(removeFruit(id))}
            value={`${id}`}
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  )
}

export default FruitItem
