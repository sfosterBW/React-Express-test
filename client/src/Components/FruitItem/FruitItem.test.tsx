import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { fruit } from '../../utils/test-helper'

import FruitItem from './FruitItem'

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}))

describe('the FruitItem component', () => {
  const component = <FruitItem fruit={fruit} />

  it('functions as expected',() => {
    const { getByTestId } = render(component)

    expect(getByTestId('item-best')).toHaveTextContent(String(fruit.best))

    if (fruit.description) {
      expect(getByTestId('item-description'))
        .toHaveTextContent(String(fruit.description))
    }

    fireEvent.click(getByTestId('update-button'))
    expect(mockDispatch.mock.calls).toHaveLength(1)

    fireEvent.click(getByTestId('remove-button'))
    expect(mockDispatch.mock.calls).toHaveLength(2)
  })

  it('renders the same as last time', () => {
    const { container } = render(component)
    expect(container).toMatchSnapshot()
  })
})
