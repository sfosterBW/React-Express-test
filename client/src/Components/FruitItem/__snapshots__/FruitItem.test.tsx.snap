// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the FruitItem component renders the same as last time 1`] = `
<div>
  <div
    class="row"
    data-testid="item"
  >
    <div
      class="header"
    >
      <div
        class="section"
      >
        <h3
          class="subtitle"
          data-testid="item-title"
        >
          apple
        </h3>
      </div>
    </div>
    <div
      class="active"
      style="height: 0px; transition: all 0.5s;"
    >
      <div
        class="sectionRow"
      >
        <h4
          class="subtitle"
        >
          Description
        </h4>
        <p
          class="description"
          data-testid="item-description"
        >
          "I don't like doctors"
        </p>
      </div>
      <div
        class="section"
      >
        <h4
          class="subtitle"
        >
          Best
        </h4>
        <input
          class="button"
          id="1"
          name="1"
          type="checkbox"
          value="1"
        />
        <label
          data-testid="item-best"
          for="1"
        >
          false
        </label>
      </div>
      <div
        class="section"
      >
        <h4
          class="subtitle"
        >
          Edit
        </h4>
        <button
          class="button"
          data-testid="update-button"
          name="modal"
        >
          Update
        </button>
        <button
          class="button"
          data-testid="remove-button"
          name="remove"
          value="1"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
</div>
`;
